python/cuml/tsa/arima.pyx

```diff
@@ -0,0 +1,901 @@
+
+        # Recall: Maximimize LL means minimize negative
+        n_llf = -(ll_f(num_batches, num_samples,
+                       order,
+                       d_y, x,
+                       trans=True, handle=handle))
+        return n_llf/(num_samples-1)
+
+    # optimized finite differencing gradient for batches
+    def gf(x):
+        """The gradient of the (batched) energy functional."""
+        # Recall: We maximize LL by minimizing -LL
+        n_gllf = -ll_gf(num_batches, num_samples,
+                        num_parameters, order,
+                        d_y, x, h,
+                        trans=True, handle=handle)
+        return n_gllf/(num_samples-1)
+
+    x0 = pack(p, d, q, num_batches, mu0, ar_params0, ma_params0)
+    x0 = _batch_invtrans(p, d, q, num_batches, x0, handle)
+
+    # check initial parameter sanity
+    if ((np.isnan(x0).any()) or (np.isinf(x0).any())):
+        raise FloatingPointError("Initial condition 'x0' has NaN or Inf.")
+
+    # Optimize parameters by minimizing log likelihood.
+    x, niter, flags = batched_fmin_lbfgs_b(f, x0, num_batches, gf,
+                                           iprint=opt_disp, factr=1000)
+
+    # Handle non-zero flags with Warning
+    if (flags != 0).any():
+        print("WARNING(`fit()`): Some batch members had optimizer problems.")
+
+    Tx = _batch_trans(p, d, q, num_batches, x, handle)
+    mu, ar, ma = unpack(p, d, q, num_batches, Tx)
+
+    fit_model = ARIMAModel(order, mu, ar, ma, y)
+    fit_model.niter = niter
+    fit_model.d_y = d_y
+    return fit_model
+
+
+def grid_search(y_b, d=1, max_p=3, max_q=3, method="bic"):
+    """Grid search to find optimal model order (p, q), weighing
+    model complexity against likelihood.
+    Optimality is based on minimizing BIC or AIC, which
+    both sum negative log-likelihood against model complexity; Higher model
+    complexity might yield a lower negative LL, but at higher `bic` due to
+    complexity term.
+
+    Parameters:
+
+    y_b : array-like shape = (n_samples, n_series)
+          The batched time-series data.
+    d : int
+        Trend (d>0) or not (d==0)
+    max_p : int
+            Maximum `p` in search
+    max_q : int
+            Maximum `q` in search
+    method : str
+             Complexity method to use ("bic" or "aic")
+
+    Returns:
+
+    Tuple of "best" order, mu, ar, and ma parameters with the
+    corresponding IC for each series.
+
+    (best_order: List[Tuple[int, int, int]],
+     best_mu: array,
+     best_ar_params: List[array],
+     best_ma_params: List[array],
+     best_ic: array)
+
+    """
+
+    num_batches = y_b.shape[1]
+    best_ic = np.full(num_batches, np.finfo(np.float64).max/2)
+
+    best_order = num_batches*[None]
+    best_mu = np.zeros(num_batches)
+    best_ar_params = num_batches*[None]
+    best_ma_params = num_batches*[None]
+
+    for p in range(0, max_p):
+        arparams = np.zeros(p)
+        for q in range(0, max_q):
+            maparams = np.zeros(q)
```

Comment:
Are `arparams` and `maparams` actually used anywhere?

Reply:
Doesn't look like it. Probably used it in a previous version and forgot to take it out.

Reply:
I removed these variables.